### PR TITLE
Keep invalid percent escapes literal in Bun.serve route params and cookies

### PR DESCRIPTION
**Pull request by robobun**
`req.params` values and cookie values (`req.cookies`, `Bun.CookieMap`) are percent-decoded by `Bun::decodeURIComponentSIMD`. For a `%` not followed by two hex digits, the decoder emitted U+FFFD and then advanced the cursor past the `%` and the two characters after it, destroying data that was never part of an escape.

### Repro

```js
using srv = Bun.serve({
  port: 0,
  routes: { "/p/:v": req => new Response(req.params.v) },
});
// GET /p/50%-off -> "50\uFFFDff"   ("-o" consumed by a bogus escape)
// GET /p/a%zzb   -> "a\uFFFDb"
// GET /p/abc%    -> "abc\uFFFD"
// GET /p/x%2     -> "x\uFFFD2"

new Bun.CookieMap("discount=50%-off").get("discount"); // "50\uFFFDff"
```

Per RFC 3986 section 2.1 (and the WHATWG URL percent-decode algorithm), a `%` not followed by two hex digits is not an escape and passes through literally. Bun's routing docs say "Invalid unicode is replaced with the unicode replacement character", which is about the decoded bytes (a lone `%FF` is not valid UTF-8), not about a `%` that never formed an escape. In practice, any route parameter or cookie value containing a literal `%` was silently corrupted, and the adjacent characters were destroyed with it.

### Cause

In `src/jsc/bindings/decodeURIComponentSIMD.cpp`, the non-hex branch appended U+FFFD and did `cursor += 3`, and the truncated-escape branch (`%` or `%X` at the end of the input) appended U+FFFD as well.

### Fix

Both branches now append the literal `%` and advance by one, so the following characters are scanned normally. Valid `%XX` escapes are unchanged, and a well-formed escape whose decoded bytes are not valid UTF-8 (a lone `%FF`, an incomplete multibyte sequence) still becomes U+FFFD as documented.

`decodeURIComponentSIMD` has exactly three callers: `ServerRouteList` (`req.params`), `CookieMap` (`req.cookies` / `Bun.CookieMap`), and the `bun:internal-for-testing` export used by its unit test. Both user-facing surfaces get the same fix.

### Verification

- `test/js/bun/http/bun-serve-routes.test.ts`: end-to-end `req.params` table. Five of the seven rows fail on the unmodified build; `%25` and a valid UTF-8 escape are positive controls.
- `test/js/bun/http/decodeURIComponentSIMD.test.ts`: the existing "edge cases" describe asserted the U+FFFD-and-consume-3 behavior (its comments call it "the cursor advancement bug"). It is rewritten to assert literal pass-through, plus new cases for truncated escapes, consecutive `%`, adjacent valid escapes, and the preserved invalid-UTF-8 to U+FFFD path.
- `test/js/bun/cookie/cookie-map.test.ts`: `Bun.CookieMap` string parsing.
- `test/js/bun/util/cookie.test.js`: the ported `cookie` npm package suite's "should return original value on escape error" (`cookie.parse("foo=%1;bar=bar")` must give `{ foo: "%1", bar: "bar" }`) was marked `it.failing` as a known compatibility gap. This change makes it pass, so the marker is removed.

All four files pass with the fix (225 / 48 / 25 / 101) and the new assertions fail without it. `bun-serve-cookies.test.ts` (25) and the rest of `test/js/bun/cookie/` (150) also pass.

The routing docs gain one sentence stating that a `%` not followed by two hex digits is preserved literally.

Related: #32822 fixes HEAD dispatch for per-method route objects; it was reported together with this but is an independent change.

Rebased onto main after #33072 (hardening round 11), which added UTF-8 byte-input handling to `decodeURIComponentSIMD`. The two edited branches were untouched upstream; the conflicts were in adjacent tests and the routing doc (whose `\uFFFD` typo #33072's fact-check pass also fixed, so that commit dropped out). I kept the new upstream tests alongside mine and added one case in the UTF-8 byte-input describe that exercises an invalid escape in non-ASCII input, covering the intersection of the two changes.

---

**[review]** gate passed · iteration 13 · 6 files touched

<details><summary>fails on main (without fix)</summary>

```console
ASAN without fix: 14 failed, 18 skipped
$ BUN_DEBUG_QUIET_LOGS=1 bun scripts/build.ts --profile=debug --quiet test "--reporter=junit" "--reporter-outfile=/tmp/mechgate.xml" test/js/bun/cookie/cookie-map.test.ts test/js/bun/http/bun-serve-routes.test.ts test/js/bun/http/decodeURIComponentSIMD.test.ts test/js/bun/util/cookie.test.js
info: syncing channel updates for nightly-2026-05-06-x86_64-unknown-linux-gnu
info: latest update on 2026-05-06 for version 1.97.0-nightly (e95e73209 2026-05-05)
info: component rust-src is up to date
info: checking for self-update (current version: 1.29.0)
bun test v1.4.0 (0e97f6b97)

test/js/bun/cookie/cookie-map.test.ts:
(pass) Bun.Cookie and Bun.CookieMap > can create a basic Cookie [36.02ms]
(pass) Bun.Cookie and Bun.CookieMap > can create a Cookie with options [8.70ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.toString() formats properly [9.22ms]
(pass) Bun.Cookie and Bun.CookieMap > can set Cookie expires as Date [5.36ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.isExpired() returns correct value [6.48ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.isExpired() gives Max-Age precedence over Expires [9.53ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.parse works wi
... (truncated)

release without fix: 33 failed, 18 skipped
bun test v1.4.0-canary.1 (1498d7b77)

test/js/bun/cookie/cookie-map.test.ts:
(pass) Bun.Cookie and Bun.CookieMap > can create a basic Cookie [0.14ms]
(pass) Bun.Cookie and Bun.CookieMap > can create a Cookie with options [0.10ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.toString() formats properly [0.15ms]
(pass) Bun.Cookie and Bun.CookieMap > can set Cookie expires as Date [34.59ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.isExpired() returns correct value [0.25ms]
93 |   test("Cookie.isExpired() gives Max-Age precedence over Expires", () => {
94 |     const past = new Date(Date.now() - 1000);
95 |     const future = new Date(Date.now() + 86_400_000);
96 | 
97 |     // Max-Age wins over Expires regardless of which one the header listed first.
98 |     expect(new Bun.Cookie("name", "value", { maxAge: 3600, expires: past }).isExpired()).toBe(false);
                                                                                              ^
error: expect(received).toBe(expected)

Expected: false
Received: true

      at <anonymous> (/workspace/bun/test/js/bun/cookie/cookie-map.test.ts:98:90)
(fail) Bun.Cookie and Bun.CookieMap > Cookie.isExpired() gives Ma
... (truncated)
```

</details>

<details><summary>passes on PR (with fix)</summary>

```console
ASAN with fix: 18 skipped
$ BUN_DEBUG_QUIET_LOGS=1 bun scripts/build.ts --profile=debug --quiet test "--reporter=junit" "--reporter-outfile=/tmp/mechgate.xml" test/js/bun/cookie/cookie-map.test.ts test/js/bun/http/bun-serve-routes.test.ts test/js/bun/http/decodeURIComponentSIMD.test.ts test/js/bun/util/cookie.test.js
info: syncing channel updates for nightly-2026-05-06-x86_64-unknown-linux-gnu
info: latest update on 2026-05-06 for version 1.97.0-nightly (e95e73209 2026-05-05)
info: component rust-src is up to date
info: checking for self-update (current version: 1.29.0)
bun test v1.4.0 (0e97f6b97)

test/js/bun/cookie/cookie-map.test.ts:
(pass) Bun.Cookie and Bun.CookieMap > can create a basic Cookie [5.35ms]
(pass) Bun.Cookie and Bun.CookieMap > can create a Cookie with options [5.63ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.toString() formats properly [5.59ms]
(pass) Bun.Cookie and Bun.CookieMap > can set Cookie expires as Date [3.43ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.isExpired() returns correct value [4.11ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.isExpired() gives Max-Age precedence over Expires [5.96ms]
(pass) Bun.Cookie and Bun.CookieMap > Cookie.parse works wit
... (truncated)

release with fix: 18 skipped
$ bun scripts/build.ts --profile=release
info: syncing channel updates for nightly-2026-05-06-x86_64-unknown-linux-gnu
info: latest update on 2026-05-06 for version 1.97.0-nightly (e95e73209 2026-05-05)
info: component rust-src is up to date
info: checking for self-update (current version: 1.29.0)
[configured] bun-profile → bun (stripped) in 719ms (unchanged)
ninja: Entering directory `/workspace/bun/build/release'
[1/7] gen cpp.rs (cppbind)
[1/7] cargo bun_bin → libbun_rust.a (--target x86_64-unknown-linux-gnu)
info: syncing channel updates for nightly-2026-05-06-x86_64-unknown-linux-gnu
info: latest update on 2026-05-06 for version 1.97.0-nightly (e95e73209 2026-05-05)
info: component rust-src is up to date
info: component rust-std is up to date

  nightly-2026-05-06-x86_64-unknown-linux-gnu unchanged - rustc 1.97.0-nightly (e95e73209 2026-05-05)

info: checking for self-update (current version: 1.29.0)
[1m[92m    Blocking[0m waiting for file lock on build directory
[1m[92m    Finished[0m `release` profile [optimized + debuginfo] target(s) in 3m 22s
[3/7] cxx obj/unified/UnifiedSource-src_jsc_bindings-5.cpp.o
[4/7] link bun-profile
[5/7] bun-profile --re
... (truncated)
```

</details>

<details><summary>diff hotspot</summary>

```
docs/runtime/http/routing.mdx                   |  2 +-
 src/jsc/bindings/decodeURIComponentSIMD.cpp     |  8 +--
 test/js/bun/cookie/cookie-map.test.ts           | 13 +++++
 test/js/bun/http/bun-serve-routes.test.ts       | 20 ++++++++
 test/js/bun/http/decodeURIComponentSIMD.test.ts | 68 ++++++++++++++-----------
 test/js/bun/util/cookie.test.js                 |  2 +-
 6 files changed, 77 insertions(+), 36 deletions(-)
```

</details>

**gate history** · 1 passed · 0 rejected · iteration 13

<details><summary>evidence per changed file</summary>

```
file                                             reads  edits  tests
docs/runtime/http/routing.mdx                        5      6     27
src/jsc/bindings/decodeURIComponentSIMD.cpp          1      1     27
test/js/bun/cookie/cookie-map.test.ts                1      1      9
test/js/bun/http/bun-serve-routes.test.ts            7      9     16
test/js/bun/http/decodeURIComponentSIMD.test.ts      4      2      8
test/js/bun/util/cookie.test.js                      2      1     10
```

</details>

### CODE REVIEW

File: docs/runtime/http/routing.mdx [hunk 1]
```diff
@@ -124,7 +124,7 @@ Bun.serve({
 });
 ```
 
-Percent-encoded route parameter values are automatically decoded. Unicode characters are supported. Invalid unicode is replaced with the unicode replacement character `&0xFFFD;`.
+Percent-encoded route parameter values are automatically decoded. Unicode characters are supported. Invalid unicode is replaced with the unicode replacement character `&0xFFFD;`. A `%` that is not followed by two hexadecimal digits is not an escape and is preserved literally.
```
[the remark this answers is not in the document]
Comment (robobun) [reply docs/runtime/http/routing.mdx]: Done in 6f845794fc: now reads `(`\uFFFD`)`, matching how `docs/runtime/s3.mdx` spells the same character.

**Comment by robobun** [comment]
Closing as part of a cleanup of stale pull requests. This PR has had no new commits since 2026-07-09, it conflicts with main, and its last CI run failed. This is not a judgment on the fix itself. If the problem still reproduces on a current build, reopen this PR after a rebase or open a new one against main.
